dockstore-integration-testing/src/test/java/io/dockstore/client/cli/BitBucketBasicIT.java

```diff
@@ -0,0 +1,467 @@
+        DockstoreTool tool = toolsApi.getContainerByToolPath(toolPath, "");
+        try {
+            toolsApi.publish(tool.getId(), CommonTestUtilities.createPublishRequest(true));
+            fail("Should not be able to publish");
+        } catch (ApiException e) {
+            assertTrue(e.getMessage().contains("Repository does not meet requirements to publish"));
+        }
+    }
+
+    /**
+     * Tests publishing tools with normal structure
+     */
+    @Test
+    public void testPublishAndUnpublishTool() {
+        publishAndUnpublishToolHelper("quay.io/dockstoretestuser/quayandgithub");
+        publishAndUnpublishToolHelper("quay.io/dockstoretestuser/quayandbitbucket");
+        publishAndUnpublishToolHelper("quay.io/dockstoretestuser/quayandgitlab");
+    }
+
+    @Test
+    public void testPublishToolEvents() {
+        Assert.assertEquals("There should be no publish events", 0, testingPostgres.getPublishEventCount());
+        Assert.assertEquals("There should be no unpublish events", 0, testingPostgres.getUnpublishEventCount());
+        publishAndUnpublishToolHelper("quay.io/dockstoretestuser/quayandgithub");
+        publishAndUnpublishToolHelper("quay.io/dockstoretestuser/quayandbitbucket");
+        publishAndUnpublishToolHelper("quay.io/dockstoretestuser/quayandgitlab");
+        final long pubEventsCount = testingPostgres.getPublishEventCount();
+        final long unpubEventsCount = testingPostgres.getUnpublishEventCount();
+        Assert.assertEquals("There should be 3 publish events", 3, pubEventsCount);
+        Assert.assertEquals("There should be 3 unpublish events", 3, unpubEventsCount);
+    }
+
+    private void publishAndUnpublishToolHelper(String toolPath) {
+        ApiClient client = getWebClient(USER_1_USERNAME, testingPostgres);
+        ContainersApi toolsApi = new ContainersApi(client);
+        DockstoreTool tool = toolsApi.getContainerByToolPath(toolPath, "");
+
+        // Publish
+        tool = toolsApi.publish(tool.getId(), CommonTestUtilities.createPublishRequest(true));
+
+        final long count = testingPostgres
+            .runSelectStatement("select count(*) from tool where name = '" + toolPath.split("/")[2] + "' and ispublished='t'", long.class);
+        Assert.assertEquals("there should be 1 registered", 1, count);
+
+        // Unpublish
+        tool = toolsApi.publish(tool.getId(), CommonTestUtilities.createPublishRequest(false));
+
+        final long count2 = testingPostgres
+            .runSelectStatement("select count(*) from tool where name = '" + toolPath.split("/")[2] + "' and ispublished='t'", long.class);
+        Assert.assertEquals("there should be 0 registered", 0, count2);
+    }
+
+    /**
+     * Tests publishing tools with non-standard structure
+     */
+    @Test
+    public void testPublishAndUnpublishAlternateStructure() {
+        publishAndUnpublishAlternateStructureHelper("dockstoretestuser", "quayandgithubalternate",
```

Comment:
Hmmm, I think some of this usage of github is fine for now, but does mean that the class should be described as depending on both github and bitbucket (and gitlab)